trinity/main.py

```diff
@@ -376,15 +379,18 @@ def display_launch_logs(chain_config: ChainConfig) -> None:
     logger.info("Trinity DEBUG log file is created at %s", str(chain_config.logfile_path))
 
 
-def run_service_until_quit(service: BaseService) -> None:
-    loop = service.get_event_loop()
-    asyncio.ensure_future(exit_on_signal(service), loop=loop)
-    asyncio.ensure_future(service.run(), loop=loop)
-    loop.run_forever()
-    loop.close()
+async def handle_networking_exit(service: BaseService,
+                                 plugin_manager: PluginManager,
+                                 endpoint: Endpoint) -> None:
+
+    async with wait_for_exit(service.get_event_loop()):
+        await service.cancel()
+        await plugin_manager.shutdown()
+        endpoint.stop()
```

Comment:
With a new `wait_for_service_exit` this could be:

```py
    async with wait_for_service_exit(service):
        await plugin_manager.shutdown()
        endpoint.stop()
```